Modules/_sqlite/cursor.c

```diff
@@ -1291,11 +1367,12 @@ pysqlite_cursor_close_impl(pysqlite_Cursor *self)
         return NULL;
     }
 
-    if (self->statement) {
-        (void)stmt_reset(self->statement);
-        Py_CLEAR(self->statement);
+    if (self->statement && stmt_reset(self->statement) != SQLITE_OK) {
+        cursor_cannot_reset_stmt_error(self, 0);
```

Comment:
Here I'm a bit unsure whether we should call `Py_CLEAR(self->statement)` after a reset failure. If we cannot reset the statement, that means we were not able to free the memory (or maybe it's something else). However, the exact return codes being returned are not documented on sqlite3's side so I can't have a finer check.

Reply:
Ok, actually I could have a double-free:

- `stmt_clear` fails but internally manages to free the memory.
- self->statement isnot cleared so still not NULL
- someone recalls close() -> we retry to clear the memory -> double free.